File: DMMPVP/core/src/main/java/com/zenyte/game/content/wildernessVault/queenreaver/BloodTrail.java
package com.zenyte.game.content.wildernessVault.queenreaver;


import com.zenyte.game.util.Utils;
import com.zenyte.game.world.World;
import com.zenyte.game.world.entity.Location;
import com.zenyte.game.world.object.WorldObject;

/**
 * @author Jire
 * @author Tommeh
 */
public class BloodTrail extends WorldObject {

	private static final int initialTicks = 30;

	private final QueenReaver queen;
	private int ticks = initialTicks;

	public BloodTrail(QueenReaver queen, Location tile) {
		super(32984, DEFAULT_TYPE, Utils.random(3), tile);
		this.queen = queen;
		this.queen.addSplat(getLocation());
		World.spawnObject(this);
	}

	public boolean process() {
		if (!queen.canProcess()) return false;

		if (--ticks == 0) {
			queen.removeSplat(getLocation());
			World.removeObject(this);
			return false;
		}

		return true;
	}

	public void resetTimer() {
		ticks = initialTicks;
	}

}

File: DMMPVP/plugins/excluded/theatreofblood/src/main/kotlin/com/zenyte/game/content/theatreofblood/room/maidenofsugadinti/object/BloodTrail.kt
package com.zenyte.game.content.theatreofblood.room.maidenofsugadinti.`object`

import com.zenyte.game.content.theatreofblood.room.maidenofsugadinti.npc.MaidenOfSugadinti
import com.zenyte.game.util.Utils
import com.zenyte.game.world.World
import com.zenyte.game.world.entity.Location
import com.zenyte.game.world.`object`.WorldObject

/**
 * @author Jire
 * @author Tommeh
 */
internal class BloodTrail(
    private val maiden: MaidenOfSugadinti,
    private val tile: Location,
    private val initialTicks: Int = 1 + (if (maiden.room.raid.hardMode) 45 else 30)
) : WorldObject(32984, DEFAULT_TYPE, Utils.random(3), tile) {

    private var ticks = initialTicks

    fun process() = if (maiden.dead()) false
    else when (--ticks) {
        initialTicks - 1 -> {
            maiden.addSplat(tile)
            World.spawnObject(this)
            true
        }
        0 -> {
            remove()
            false
        }
        else -> true
    }

    fun remove() {
        maiden.removeSplat(tile)
        World.removeObject(this)
    }

    fun resetTimer() {
        ticks = initialTicks
    }

}